plasma_framework/contracts/poc/fast_exits/Quasar.sol

```diff
@@ -0,0 +1,335 @@
+        require(outputOwner == expectedOutputOwner, "Was not called by the Output owner");
+    }
+
+    /**
+     * @dev Verify the challengeTx spends the output
+     * @param utxoPos pos of the output, which is the ticket identifier
+     * @param rlpChallengeTx RLP-encoded challenge transaction
+     * @param challengeTxInputIndex index pos of the same utxo in the challenge transaction
+     * @param challengeTxWitness Witness for challenging transaction
+    */
+    function verifySpendingCondition(uint256 utxoPos, bytes memory rlpChallengeTx, uint16 challengeTxInputIndex, bytes memory challengeTxWitness) private {
+        GenericTransaction.Transaction memory challengingTx = GenericTransaction.decode(rlpChallengeTx);
+
+        GenericTransaction.Transaction memory inputTx = GenericTransaction.decode(ticketData[utxoPos].rlpOutputCreationTx);
+        PosLib.Position memory utxoPosDecoded = PosLib.decode(utxoPos);
+        GenericTransaction.Output memory output = GenericTransaction.getOutput(inputTx, utxoPosDecoded.outputIndex);
+
+        ISpendingCondition condition = spendingConditionRegistry.spendingConditions(
+            output.outputType, challengingTx.txType
+        );
+        require(address(condition) != address(0), "Spending condition contract not found");
+        bool isSpent = condition.verify(
+            ticketData[utxoPos].rlpOutputCreationTx,
+            utxoPos,
+            rlpChallengeTx,
+            challengeTxInputIndex,
+            challengeTxWitness
+        );
+        require(isSpent, "Spending condition failed");
+    }
+
+    /**
+     * @dev Verify the validity of the ticket
+     * @param utxoPos pos of the output, which is the ticket identifier
+    */
+    function verifyTicketValidityForClaim(uint256 utxoPos) private {
+        require(!ticketData[utxoPos].isClaimed, "Already Claimed");
+        require(ticketData[utxoPos].outputOwner == msg.sender, "Not called by the ticket owner");
+        uint256 expiryTimestamp = ticketData[utxoPos].validityTimestamp;
+        require(expiryTimestamp != 0 && block.timestamp <= expiryTimestamp, "Ticket is not valid");
+    }
+    
+    /**
+     * @dev Verify the claim Tx is properly formed
+     * @param utxoPos pos of the output, which is the ticket identifier
+     * @param claimTx the Claim Tx to the quasar owner
+    */
+    function verifyClaimTxCorrectlyFormed(uint256 utxoPos, bytes memory claimTx) private {
+        PaymentTransactionModel.Transaction memory decodedTx
+        = PaymentTransactionModel.decode(claimTx);
+
+        // first input should be the Utxo with which ticket was obtained
+        require(decodedTx.inputs[0] == bytes32(utxoPos), "Wrong Tx provided");
+
+        // first output should be the utxo for Quasar owner
+        FungibleTokenOutputModel.Output memory outputData
+        = PaymentTransactionModel.getOutput(decodedTx, 0);
+
+        // verify output to Quasar Owner
+        verifyOwnership(outputData, quasarOwner);
```

Comment:
The error message if this fails doesn't make sense - see comment on `verifyOwnership()` above